internal/feeder/bastion/bastion_feeder.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bastion is an implementation of a witness feeder which talks to a bastion server.
+package bastion
+
+import (
+	"bufio"
+	"context"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/transparency-dev/formats/log"
+	"github.com/transparency-dev/witness/internal/config"
+	"github.com/transparency-dev/witness/internal/feeder"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+)
+
+type Config struct {
+	Addr            string
+	Prefix          string
+	Logs            []config.Log
+	BastionKey      ed25519.PrivateKey
+	WitnessVerifier note.Verifier
+}
+
+// FeedBastion talks to the bastion to receive checkpoints to be witnessed.
+// This function returns once the provided context is done.
+func FeedBastion(ctx context.Context, c Config, w feeder.Witness) error {
+	klog.Infof("My bastion backend ID: %064x", sha256.Sum256(c.BastionKey.Public().(ed25519.PublicKey)))
+	h := &addHandler{
+		w:           w,
+		logs:        make(map[string]config.Log),
+		witVerifier: c.WitnessVerifier,
+	}
+	for _, l := range c.Logs {
+		l := l
+		h.logs[l.ID] = l
+	}
+
+	return connectAndServe(ctx, c.Addr, h, c.BastionKey)
+}
+
+type addHandler struct {
+	w           feeder.Witness
+	logs        map[string]config.Log
+	witVerifier note.Verifier
+}
+
+func (a *addHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	_, proof, cp, err := parseBody(r.Body)
+	if err != nil {
+		klog.V(1).Infof("invalid body: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	s := strings.SplitN(string(cp), "\n", 2)
+	if len(s) != 2 {
```

Comment:
Won't this break if the checkpoint has extension lines?

Reply:
No, because it's produced with `strings.SplitN(_, _, 2)`